### PR TITLE
Gradient definitions & supertypes

### PR DESCRIPTION
**Pull request by mcabbott**
This is inspired by https://github.com/SciML/DiffEqSensitivity.jl/pull/493#issuecomment-927014502 . I wrote many examples there, but the core surprise was that a `VectorOfArrays{T} <: AbstractArray{T,3}` has a gradient `Vector{Matrix{T}}` which is not a 3-array. If I understood correctly, it would desirable that the result be another VectorOfArray, i.e. something `AbstractArray{T,3}` which iterates like Vector{Matrix}.

One way to do that would be using ProjectTo. But in fact I believe the Vector{Matrix} is being explicitly generated by a rule defined here, so perhaps it is simpler just to alter that rule to wrap this?

RFC, WIP, needs tests... and needs thought about other cases:
```julia
julia> va = RecursiveArrayTools.VectorOfArray([rand(3,3), rand(3,3)]);

julia> gradient(va -> sum(va[1]), va)[1]
VectorOfArray{Float64,3}:
2-element Vector{AbstractMatrix{Float64}}:
 3×3 Fill{Float64}, with entries equal to 1.0
 [0.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0]

julia> ans + va |> summary  # essential this work for a valid tangent
"VectorOfArray{Float64,3}"

julia> gradient((x,y) -> sum(x ./ y), va, rand(3))[1] |> summary  # should this also be changed?
"3×3×2 Array{Float64, 3}"

julia> gradient(va -> sum(va[:,1]), va)[1]  # this didn't work before, either. 
ERROR: BoundsError: attempt to access 3×3×2 VectorOfArray{Float64, 3, Vector{Matrix{Float64}}} at index [1:3, 1]
```
* The broadcast case could apply something like `VectorOfArray(collect(eachslice(Δ)))`, to return a comparable object, i.e. one which broadcasts one way and iterates a different way. This would need `ProjectTo`. It would also apply to things like `gradient(va -> sum(abs, [1 2; 3 4] * va), VectorOfArray([[1,2], [3,4]]))`. 
* Is indexing `va[:,1]` expected to work? What should it return? The other way around should I think do this:
```
julia> gradient(va -> sum(va[1,:]), va)[1]
VectorOfArray{Float64,3}:
2-element Vector{Matrix{Float64}}:
 [1.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0]
 [1.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0]
```

### CODE REVIEW

File: src/zygote.jl [hunk 1]
```diff
@@ -55,11 +73,13 @@ ZygoteRules.@adjoint function getindex(VA::AbstractVectorOfArray, i::Union{Int,A
   function AbstractVectorOfArray_getindex_adjoint(Δ)
     Δ′ = zero(VA)
     Δ′[i,j...] = Δ
-    (Δ′, i,map(_ -> nothing, j)...)
+    @show Δ′
+    # (Δ′, i,map(_ -> nothing, j)...)  # surely that i is a bug?
+    (Δ′, nothing, map(_ -> nothing, j)...)
+    # (VectorOfArray(Δ′), nothing, map(_ -> nothing, j)...)
```
Comment (ChrisRackauckas) [inline src/zygote.jl:79]: why would it not be this one?

File: src/zygote.jl [hunk 2]
```diff
@@ -55,11 +73,13 @@ ZygoteRules.@adjoint function getindex(VA::AbstractVectorOfArray, i::Union{Int,A
   function AbstractVectorOfArray_getindex_adjoint(Δ)
     Δ′ = zero(VA)
     Δ′[i,j...] = Δ
-    (Δ′, i,map(_ -> nothing, j)...)
+    @show Δ′
```
Comment (ChrisRackauckas) [inline src/zygote.jl:76]:
```suggestion
```

**Comment by ChrisRackauckas** [comment]
Thanks! Running downstream tests to check it out.

File: src/zygote.jl (hunk 2 above)
Comment (mcabbott) [reply src/zygote.jl:76]: Sorry, very WIP!

File: src/zygote.jl [hunk 3]
```diff
@@ -55,11 +73,13 @@ ZygoteRules.@adjoint function getindex(VA::AbstractVectorOfArray, i::Union{Int,A
   function AbstractVectorOfArray_getindex_adjoint(Δ)
     Δ′ = zero(VA)
     Δ′[i,j...] = Δ
-    (Δ′, i,map(_ -> nothing, j)...)
+    @show Δ′
+    # (Δ′, i,map(_ -> nothing, j)...)  # surely that i is a bug?
```
Comment (ChrisRackauckas) [inline src/zygote.jl:77]: yeah that's a bug

File: src/zygote.jl (hunk 2 above)
Comment (ChrisRackauckas) [reply src/zygote.jl:76]: oh no worries, still worth the downstream tests

**Pull request merged**